proposals/TP1-xds-transport-next.md

```diff
@@ -0,0 +1,1028 @@
+protocol and does not require any changes to clients or servers that do not wish
+to opt into the new naming scheme.
+
+### URI-based xDS resource names
+
+The following URI schema is proposed for xDS resource names:
+
+`xdstp://[{authority}]/{resource type}/{id/*}?{context parameters}{#processing directive,*}`
+
+* `authority` is an opaque string naming a resource authority, e.g.
+  `some.control.plane`.
+* `resource type` is the xDS resource type URL, e.g. `envoy.config.route.v3.RouteConfiguration`
+* `id/*` is the remaining path component of the URI and is fully opaque; naming
+  is at the discretion of the control plane(s).
+* `context parameters` are the URI query parameters and express contextual
+  information for selecting resource variants, for example `shard_id=1234&direction=inbound`.
+* A `processing directive` provides additional information to the client on how
+  the URI is to be interpreted. Two supported directives are:
+  * `alt=<xdstp:// URI>` expressing a fallback URI.
+  * `entry=<resource name>` providing an anchor referring to an inlined resource
+    for [list collections](#list). Resource names must be of the form `[a-zA-Z0-9_-\./]+`.
+
+Multiple directives are comma separated in the fragment component of the URI.
+
+An example application of the above URI scheme is:
+
+`xdstp://some.control.plane/envoy.config.route.v3.RouteConfiguration/foo/bar?shard_id=1234&direction=inbound`
+
+We distinguish between two forms of `xdstp://` URIs:
+
+* Uniform Resource Locators (URLs) instructing a client on how to locate a given
+  xDS resource. URLs may include processing directives.
+
+* Uniform Resource Names (URNs) referring to a specific xDS resource. URNs do not
+  include processing directives. `xdstp://` URNs provide self-contained cache keys.
+
+#### Proto3 expression
+
+`xdstp://` URIs are canonically encoded in `proto3` messages. The structure
+for this encoding is provided at https://github.com/cncf/xds/tree/main/xds/core/v3:
+
+* [URL](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_locator.proto) definition
+* [URN](https://github.com/cncf/xds/blob/main/xds/core/v3/resource_name.proto) definition
+
+#### Normalization
+
+Two `xdstp://` URNs are considered equivalent if they match component-wise
+modulo context parameter ordering.
+
+## Context parameters
+
+Context parameters in URNs presented by the client to the server will be dervied
+by the following process. Using an example of a URL
+`xdstp://some-authority/some.type/foo?xds.bar=baz`:
+
+1. Static `Node`-derived context parameters are first added. These are prefixed with
+   `xds.node.`. The set of `Node`-derived context parameters is specified in the
+   bootstrap on a per-resource type basis. The key reflects the `Node` proto3
+   field structure, e.g. `xds.node.locality.sub_zone=some_sub_zone`,
+   `xds.node.metadata.bar="a"`, `xds.node.user_agent_version=1.2.3`.
+
+2. Dynamic context parameters are added next. These are free form and at the
```

Comment:
I think we need to be clear on the relative ordering, since they are somehow combined with URL-derived context parameters and we need to be clear which way the override relationship works. Since we have dynamic parameters that can come from the binary, as well as the transport negotiation that you are working on, it probably makes sense to touch on them here.

Reply:
I'm going to drop both DCPs _and_ client feature capabilities from this xRFC and leave them for your dynamic context params extensions.